comp_sci_gender_bias/pipeline/glove_differences/process_text_utils.py

```diff
@@ -0,0 +1,325 @@
+    ):
+        self.glove_path = os.environ.get("GLOVE_PATH")
+        self.masc_comparisons = masc_comparisons
+        self.fem_comparisons = fem_comparisons
+        self.glove_d = glove_d  # 50, 100, 200, 300
+        self.glove_txt_file = f"glove.6B.{self.glove_d}d.txt"
+
+    def load_glove2word2vec(self):
+        output_file = os.path.join(self.glove_path, "gensim_" + self.glove_txt_file)
+        if not os.path.exists(output_file):
+            print("Gensim file being prepared...")
+            input_file = os.path.join(self.glove_path, self.glove_txt_file)
+            glove2word2vec(input_file, word2vec_output_file=output_file)
+
+        self.model = KeyedVectors.load_word2vec_format(output_file, binary=False)
+
+    def gender_similarity_difference_word_list(self, word_list):
+        """
+        Input a word or a list of words and the output will be a dictionary
+        of the masculine - feminine cosine similarity averages for each word.
+
+        Not all words will be in the corpus, so only words found will be in the output dictionary.
+        """
+        if not isinstance(word_list, list):
+            word_list = [word_list]
+
+        # There will be an error if the word isn't in the vocab
+        # also all words need to be lower case to be found in vocab
+        word_list = [
+            word.lower() for word in word_list if word in self.model.index_to_key
+        ]
+
+        if word_list:
+            word_vecs = self.model[word_list]
+
+            masc_av_similarities = np.array(
+                [
+                    self.model.cosine_similarities(self.model[word], word_vecs)
+                    for word in self.masc_comparisons
+                ]
+            ).mean(axis=0)
+            fem_av_similarities = np.array(
+                [
+                    self.model.cosine_similarities(self.model[word], word_vecs)
+                    for word in self.fem_comparisons
+                ]
+            ).mean(axis=0)
+
+            return dict(
+                zip(word_list, list(masc_av_similarities - fem_av_similarities))
+            )
+        else:
+            return None
+
+
+def get_word_freq(word_pos_df: pd.DataFrame, divide_by_pos_freq: bool = False) -> dict:
+    """
+    Get the word frequencies for a corpus.
+
+    Args:
+        word_pos_df: DataFrame with 2 columns ["Word", "POS"].
+            Each row is every word in the corpus along with it's POS tag
+        divide_by_pos_freq: Whether to divide by frequency of the specific
+            part of speech type in all corpus.
+
+    Returns:
+        A dictionary of the frequency of words in the corpus.
+        If divide_by_pos_freq is True, frequency is divided by
+        frequency of the specific part of speech type in all corpus.
+    """
+
+    pos_freq = (word_pos_df.groupby("POS").count() / len(word_pos_df))["Word"].to_dict()
+    word_pos = (
+        word_pos_df.groupby("Word")["POS"].agg(lambda x: pd.Series.mode(x)[0])
```

Comment:
possible source of discrepancy: what happens if a word has 2 POS types with equal frequencies

Reply:
a way to mitigate this could be to allow multiple records for identical words, but differentiated by POS